### PR TITLE
Websocket insight client

### PR DESCRIPTION
**Pull request by dsnipe**
Not as beauty as I want, but it works.

### CODE REVIEW

File: lib/straight-server.rb [hunk 1]
```diff
@@ -25,6 +26,7 @@ class << self
 require_relative 'straight-server/utils/hash_string_to_sym_keys'
 require_relative 'straight-server/random_string'
 require_relative 'straight-server/config'
+require_relative 'straight-server/websocket_insight_client.rb'
```
Comment (snitko) [inline lib/straight-server.rb]: Why `.rb`?

File: spec/.straight/config.yml [hunk 2]
```diff
@@ -45,6 +45,8 @@ blockchain_adapters:
 
 insight_url: "https://insight.mycelium.com/api"
 insight_test_url: "https://test-insight.bitpay.com"
+
+insight_twebsocket_url: "wss://insight.mycelium.com"
```
Comment (snitko) [inline spec/.straight/config.yml]: What's `insight_twebsocket_url`? I mean, why `twebsocket` and not just `socket`? If there is a reason and it's not a typo, an explanatory comment would be helpful.

File: spec/lib/websocket_insight_client_spec.rb [hunk 3]
```diff
@@ -0,0 +1,36 @@
+require 'spec_helper'
+
+RSpec.describe StraightServer::WebsocketInsightClient do
+
+  let(:address) {  "mttHG8y4rDrGCXhzSo5udNz8cuDRUSXSZL" }
```
Comment (snitko) [inline spec/lib/websocket_insight_client_spec.rb]: Extra space :)

File: spec/lib/websocket_insight_client_spec.rb [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+  before(:each) do
+    StraightServer::WebsocketInsightClient.clear_address_list 
+    StraightServer::WebsocketInsightClient.add_address(address)
+  end
+
+  it "add address for monitoring" do
```
Comment (snitko) [inline spec/lib/websocket_insight_client_spec.rb]: addS

File: spec/lib/websocket_insight_client_spec.rb [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+    expect(StraightServer::WebsocketInsightClient.address_for_check_list).to eq [address]
+  end
+
+  it "remove address from monitorin array" do
```
Comment (snitko) [inline spec/lib/websocket_insight_client_spec.rb]: removeS

File: spec/lib/websocket_insight_client_spec.rb [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+    StraightServer::WebsocketInsightClient.remove_address(address)
+    expect(StraightServer::WebsocketInsightClient.address_for_check_list).to eq([])
+  end
+
+  it "not add same address for monitorin array" do
```
Comment (snitko) [inline spec/lib/websocket_insight_client_spec.rb]: it doesn't add the same address

File: spec/lib/websocket_insight_client_spec.rb [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+    StraightServer::WebsocketInsightClient.add_address(address)
+    expect(StraightServer::WebsocketInsightClient.address_for_check_list).to eq([address])
+  end
+
+  it "fill order data if transaction found for specific address" do
```
Comment (snitko) [inline spec/lib/websocket_insight_client_spec.rb]: fillS order data if transaction was found for a specific address

File: templates/config.yml [hunk 8]
```diff
@@ -93,6 +93,9 @@ blockchain_adapters:
 # You can also set a testnet insight url to use:
 #   insight_test_url: "https://test.localhost/api"
 
+# Websocket Insight Url
+insight_websocket_url: "wss://insight.mycelium.com"
```
Comment (snitko) [inline templates/config.yml]: Perhaps it's not such a great idea to add this to config as a default option. Let's use Bitpay's one. We don't want all people using straight to also abuse our Insight (same goes for `spec/.straight/config.yml`)

File: lib/straight-server/order.rb [hunk 9]
```diff
@@ -60,6 +60,22 @@ def status(as_sym: false, reload: false)
       self[:status] = @status
     end
 
+    def self.set_status_for(address, data)
+      order = self.find(address: address)
```
Comment (AlexanderPavlenko) [inline lib/straight-server/order.rb]: What if it's a reused address? Incorrect order may be selected.

File: spec/.straight/config.yml (hunk 2 above)
Comment (dsnipe) [inline spec/.straight/config.yml]: it's typo. 
correct should be `insight_websocket_url`
`websocket` in naming is more concrete and explicit than just `socket`

File: lib/straight-server/websocket_insight_client.rb [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+          address = @@address_check_list.delete_at(index)
+          StraightServer.logger.info "[WS] Found transaction for address: #{address}"
+          StraightServer::Order.set_status_for(address, data)
+        end
+      end
+    end
+    
+  end
+
+end
```
Comment (AlexanderPavlenko) [inline lib/straight-server/websocket_insight_client.rb]: Well, that's quite a mess. It looks not thread-safe and higly coupled with the StraightServer::Order.
I suggest to:
1. not use class methods and class variables; instead just instantiate it during initialization and store in `StraightServer.insight_client`
2. provide an instance method `monitor(address, &block)`, which adds `address` to `@monitor` `Hash` as a key, and `block` as a value. When transaction detected, it calls that block with a `Struct` argument with all those fields https://github.com/MyceliumGear/straight-server/pull/6/files#diff-62c9cb9bdf687d36a4fa3c113e0af44eR70 (where they are assigned to Order, or whatever), and calls `stop_monitor(address)`, which deletes `address` key from `@monitor`.

File: templates/config.yml (hunk 8 above)
Comment (dsnipe) [inline templates/config.yml]: Abuse Bitpay server by default it's too not a great idea I think :)
Better I put `localhost` there and comment about it. Users will be decide which server use.

File: lib/straight-server/websocket_insight_client.rb (hunk 10 above)
Comment (dsnipe) [inline lib/straight-server/websocket_insight_client.rb]: 1. Agreed with it. 
2. It's possible to decouple it using callback (your suggestion) or pub/sub. First option is more possible in current architecture, but I don't  feel it's right (create more complexity and decrease readability)

Comment (AlexanderPavlenko) [inline lib/straight-server/websocket_insight_client.rb]: 2. What do you mean by "pub/sub"? It's a callback too.

Comment (dsnipe) [inline lib/straight-server/websocket_insight_client.rb]: Not exactly. Callback you pass as argument (in that case, add new address for monitoring). And then when some event happens (for example transaction detected) you call it. Which mean – who asked, that receives.
In pu/sub pattern, when event happens it just broadcast about it. And those who listen for that event, fire their methods. So no need to pass any additional data to event emiter and you can add as much listeners as you want in any places of application. 
Very good implementation of it is Whisper gem.
Here nice article (but related mostly to RoR): http://www.toptal.com/ruby-on-rails/the-publish-subscribe-pattern-on-rails

Comment (AlexanderPavlenko) [inline lib/straight-server/websocket_insight_client.rb]: > you can add as much listeners as you want in any places of application
YAGNI. We need only one listener as an entry-point from where any handlers are called.

Comment (dsnipe) [inline lib/straight-server/websocket_insight_client.rb]: Not exactly. Here we have situation with 3 concurrent jobs (websocket and 2 parallel api requests). Who first – winner.
Each job can broadcast same event about success and we can listen in one place (and for example in another place listen about error for log or restart or whenever) and then cancel other jobs. It's only one scenario. Another – we can make websocket service on another language (who knows what will be in future) and in ruby just receive event and broadcast about it. 
So pu/sub give us very good flexibility. And imho it's much easier to read (all related logic in one place) than tracking callbacks.

Comment (AlexanderPavlenko) [inline lib/straight-server/websocket_insight_client.rb]: That's a different story. I had no assumption about Celluloid usage.

**Pull request merged**